src/librustc_mir/transform/const_prop.rs

```diff
@@ -908,17 +954,31 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             if let Ok(place_layout) = self.tcx.layout_of(self.param_env.and(place_ty)) {
                 let can_const_prop = self.can_const_prop[place.local];
                 if let Some(()) = self.const_prop(rval, place_layout, source_info, place) {
-                    // This will return None if the above `const_prop` invocation only "wrote" a
-                    // type whose creation requires no write. E.g. a generator whose initial state
-                    // consists solely of uninitialized memory (so it doesn't capture any locals).
-                    if let Some(value) = self.get_const(place) {
-                        if self.should_const_prop(value) {
-                            trace!("replacing {:?} with {:?}", rval, value);
-                            self.replace_with_const(rval, value, source_info);
-                            if can_const_prop == ConstPropMode::FullConstProp
-                                || can_const_prop == ConstPropMode::OnlyInsideOwnBlock
-                            {
-                                trace!("propagated into {:?}", place);
+                    if can_const_prop != ConstPropMode::NoPropagation {
+                        // This will return None for variables that are from other blocks,
+                        // so it should be okay to propagate from here on down.
+                        if let Some(value) = self.get_const(place) {
+                            if self.should_const_prop(value) {
+                                trace!("replacing {:?} with {:?}", rval, value);
+                                self.replace_with_const(rval, value, statement.source_info);
+                                if can_const_prop == ConstPropMode::FullConstProp
+                                    || can_const_prop == ConstPropMode::OnlyInsideOwnBlock
+                                {
+                                    trace!("propagated into {:?}", place);
+
+                                    if place.projection.is_empty()
+                                        && !self.has_mutation.contains(place.local)
+                                    {
+                                        self.fully_propagated_locals[place.local] = Some(value);
```

Comment:
whoa, that's a subtle introduction. Multiple things:

* document the field
* document the fact that it doesn't work for aggegates initialized element wise (so after https://github.com/rust-lang/rust/pull/73656)
* This is not relevant for correctness as far as I can tell, it just influences debug info, document this so if that changes we take a closer look at the effects of trusting it for correctness.

Reply:
I couldn't find any tests actually using this on aggegates... Don't we have any or is there another reason?

Reply:
I'm personally not a huge fan of having this field. It seems easy to add logic later that uses it incorrectly. 

Here's a different idea, I think we could remove this field entirely and instead loop over the debuginfo here and update it accordingly. @oli-obk do you think that would be better?

Reply:
Another alternative is to make it a separate mir optimization that goes through and looks for locals that are assigned exactly once and directly from an `Rvalue::Operand(Operand::Const(_))` and just move that constant into the debug info.

Reply:
> Another alternative is to make it a separate mir optimization

I did this and I like the result a lot better.